tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphFunction.java

```diff
@@ -0,0 +1,301 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_FunctionName;
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_FunctionSetAttrValueProto;
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_GraphToFunction;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.bytedeco.javacpp.BytePointer;
+import org.bytedeco.javacpp.Pointer;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.Graph.Reference;
+import org.tensorflow.internal.c_api.TF_Function;
+import org.tensorflow.internal.c_api.TF_Operation;
+import org.tensorflow.internal.c_api.TF_Output;
+import org.tensorflow.internal.c_api.TF_Status;
+import org.tensorflow.op.Ops;
+import org.tensorflow.proto.framework.AttrValue;
+import org.tensorflow.proto.framework.TensorInfo;
+
+/**
+ * A Function created from a TensorFlow {@link Graph}. Callable in Graph or Eager mode via the {@link
+ * Ops#callFunction(GraphFunction, Operand[])} ops or the {@link #call(Ops, Operand[])} methods.
+ *
+ * <p>By default, resolves arguments and returns outputs by position.
+ * Names can be specified for the inputs and outputs (and used when calling) using {@link NamedGraphFunction} and {@link
+ * #create(Graph, String, boolean, Map, Map, String)}.
+ *
+ * Functions can also be created from {@link ConcreteFunction}s using {@link #create(ConcreteFunction, boolean)}.
+ */
+public class GraphFunction implements AutoCloseable {
```

Comment:
`GraphFunction` is a bit confusing as a name, it sounds like it is a function attached to a `Graph` which is not necessarily a case. I'd turn this over to `FunctionGraph`.

Reply:
I agree, though lets have the "merge with ConcreteFunction" discussion first.  I didn't want to use `FunctionGraph` because Python has a `FuncGraph` that subclasses Graph, and I'm not sure if we'll end up needing one for further Function support (I'd prefer a subclass of Ops, which should handle most of the same stuff).